torch/utils/data/datapipes/_typing.py

```diff
@@ -477,14 +477,19 @@ def __init__(self, iterator, source_dp, iterator_id):
             self.iterator = iterator
             self.source_dp = source_dp
             self.iterator_id = iterator_id
+            self._profiler_enabled = torch.autograd._profiler_enabled()
```

Comment:
Yeah. I think there should be a special case here.
```py
it = iter(dp)  # _profiler_enabled = False

with profiler.profiler() as prof:
    for d in it:  # Profiler should be enabled but not
       ...
```

Reply:
Sure, it doesn't work if you start the iteration of DataPipe outside of the profiler context:
```python
import torch
from torchdata.datapipes.iter import IterableWrapper

dp = IterableWrapper(range(5))
it = iter(dp)
print(next(it))
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    print(next(it))
```
It will return this:
```
0
1
```
But if you start the iteration inside:
```python
dp = IterableWrapper(range(5))
it = iter(dp)
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    print(next(it))
    print(next(it))
```
It returns:
```
ENABLED
0
ENABLED
1
```

In the most common use case, it should look like:
```python
dp = IterableWrapper(range(5))
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    it = iter(dp)
    print(next(it))
    print(next(it))
```
It returns:
```
ENABLED
0
ENABLED
1
```

Reply:
@ejguan That's what I expected too, but I just tested it by modifying and got the above result. 
```python
def profiler_record_fn_context():
    print("ENABLED")
    return torch.autograd.profiler.record_function(profile_name)
```

Reply:
The only reason I can think of is when this line is executed for every `__next__` but I doubt it 

https://github.com/pytorch/pytorch/blob/3415a4c663f1ed8dc2b0f3622066782b7731ac1d/torch/utils/data/datapipes/_typing.py#L508

Reply:
Yeah

Reply:
If we worry about users creating the iterator  outside of the profiler (not the DataPipe, creating that outside is fine) , then we can run `torch.autograd._profiler_enabled()` at every `__next__`.